File: src/foam/swift/dao/AbstractDAO.js
/**
 * @license
 * Copyright 2016 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

foam.CLASS({
  package: 'foam.swift.dao',
  name: 'AbstractDAO',
  implements: [
    'foam.dao.DAO',
  ],
  requires: [
    'foam.dao.ResetListener',
    'foam.dao.LimitedSink',
    'foam.dao.OrderedSink',
    'foam.dao.PredicatedSink',
    'foam.dao.SkipSink',
  ],
  properties: [
    {
      name: 'of',
      swiftType: 'ClassInfo',
    },
    {
      name: 'primaryKey',
      swiftType: 'PropertyInfo',
      swiftExpressionArgs: ['of'],
      swiftExpression: 'return of.axiom(byName: "id") as! PropertyInfo',
    },
  ],
  methods: [
    {
      name: 'put',
      swiftCode: 'fatalError()',
    },
    {
      name: 'remove',
      swiftCode: 'fatalError()',
    },
    {
      name: 'find',
      swiftCode: 'fatalError()',
    },
    {
      name: 'select',
      swiftCode: 'fatalError()',
    },
    {
      name: 'removeAll',
      swiftCode: 'fatalError()'
    },
    {
      name: 'pipe',
      swiftCode: 'fatalError()'
    },
    {
      name: 'where',
      swiftCode: 'fatalError()'
    },
    {
      name: 'orderBy',
      swiftCode: 'fatalError()'
    },
    {
      name: 'skip',
      swiftCode: 'fatalError()'
    },
    {
      name: 'limit',
      swiftCode: 'fatalError()'
    },
    {
      name: 'decorateSink_',
      args: [
        {
          name: 'sink',
          swiftType: 'Sink',
        },
        {
          name: 'skip',
          swiftType: 'Int?',
        },
        {
          name: 'limit',
          swiftType: 'Int?',
        },
        {
          name: 'order',
          swiftType: 'Comparator?',
        },
        {
          name: 'predicate',
          swiftType: 'FoamPredicate?',
        },
      ],
      swiftReturnType: 'Sink',
      swiftCode: function() {/*
var sink = sink
if limit != nil {
  sink = LimitedSink_create([
    "limit": limit,
    "delegate": sink
  ])
}
if skip != nil {
  sink = SkipSink_create([
    "skip": skip,
    "delegate": sink
  ])
}
if order != nil {
  sink = OrderedSink_create([
    "comparator": order,
    "delegate": sink,
  ])
}
if predicate != nil {
  sink = PredicatedSink_create([
    "predicate": predicate,
    "delegate": sink,
  ])
}
return sink
      */},
    },
    {
      name: 'decorateListener_',
      args: [
        {
          name: 'sink',
          swiftType: 'Sink',
        },
        {
          name: 'skip',
          swiftType: 'Int?',
        },
        {
          name: 'limit',
          swiftType: 'Int?',
        },
        {
          name: 'order',
          swiftType: 'Comparator?',
        },
        {
          name: 'predicate',
          swiftType: 'FoamPredicate?',
        },
      ],
      swiftReturnType: 'Sink',
      swiftCode: function() {/*
// TODO: There are probably optimizations we can make here
// but every time I try it comes out broken.  So for the time being,
// if you have any sort of skip/limit/order/predicate we will just
// issue reset events for everything.
if skip != nil || limit != nil || order != nil || predicate != nil {
  return self.ResetListener_create(["delegate": sink])
}
return sink
      */},
    },
    {
      name: 'listen',
      swiftCode: function() {/*
let mySink = decorateListener_(sink, skip, limit, order, predicate)
return sub(listener: { (sub: Subscription, args: [Any?]) -> Void in
  guard let topic = args[0] as? String else { return }
  switch topic {
    case "put":
      mySink.put(sub, args.last as! FObject)
      break
    case "remove":
      mySink.remove(sub, args.last as! FObject)
      break
    case "reset":
      mySink.reset(sub)
      break
    default:
      break
  }
})
      */},
    },
  ]
});

File: src/foam/dao/AbstractDAO.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.dao;

import foam.core.*;
import foam.dao.*;
import foam.mlang.*;
import foam.mlang.predicate.*;
import foam.mlang.order.*;

public abstract class AbstractDAO
  extends    ContextAwareSupport
  implements DAO
{
  public final static long MAX_SAFE_INTEGER = 9007199254740991l;

  protected ClassInfo    of_          = null;
  protected PropertyInfo primaryKey_  = null;

  public DAO where(Predicate predicate) {
    return new FilteredDAO(predicate, this);
  }

  public DAO orderBy(Comparator comparator) {
    return new OrderedDAO(comparator, this);
  }

  public DAO skip(long count) {
    return new SkipDAO(count, this);
  }

  public DAO limit(long count) {
    return new LimitedDAO(count, this);
  }

  public void pipe_(X x, foam.dao.Sink sink) {
    throw new UnsupportedOperationException();
  }

  protected Sink decorateListener_(Sink sink, Predicate predicate) {
    if ( predicate != null ) {
      sink = new PredicatedSink(predicate, sink);
    }

    return sink;
  }

  protected Sink decorateSink_(Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
    if ( ( limit > 0 ) && ( limit < this.MAX_SAFE_INTEGER ) ) {
      sink = new LimitedSink(limit, 0, sink);
    }

    if ( ( skip > 0 ) && ( skip < this.MAX_SAFE_INTEGER ) ) {
      sink = new SkipSink(skip, 0, sink);
    }

    if ( order != null ) {
      sink = new OrderedSink(order, null, sink);
    }

    if ( predicate != null ) {
      sink = new PredicatedSink(predicate, sink);
    }

    return sink;
  }

  public ClassInfo getOf() {
    return of_;
  }

  public AbstractDAO setOf(ClassInfo of) {
    of_ = of;
    primaryKey_ = (PropertyInfo) of.getAxiomByName("id");
    return this;
  }

  public PropertyInfo getPrimaryKey() {
    return primaryKey_;
  }

  protected Object getPK(FObject obj) {
    return getPrimaryKey().get(obj);
  }

  protected class DAOListener implements foam.core.Detachable {
    protected Sink sink;
    protected java.util.Collection listeners;

    public DAOListener(Sink sink, java.util.Collection listeners) {
      this.sink = sink;
      this.listeners = listeners;
    }

    public void detach() {
      listeners.remove(this);
    }

    public void put(FObject obj) {
      try {
        sink.put(obj, this);
      } catch (java.lang.Exception e) {
        detach();
      }
    }

    public void remove(FObject obj) {
      try {
        sink.remove(obj, this);
      } catch (java.lang.Exception e) {
        detach();
      }
    }

    public void reset() {
      try {
        sink.reset(this);
      } catch (java.lang.Exception e) {
        detach();
      }
    }
  }

  protected java.util.List<DAOListener> listeners_ = new java.util.concurrent.CopyOnWriteArrayList<DAOListener>();

  public void listen(Sink sink, Predicate predicate) {
    this.listen_(this.getX(), sink, predicate);
  }

  public void listen_(X x, Sink sink, Predicate predicate) {
    sink = decorateListener_(sink, predicate);
    listeners_.add(new DAOListener(sink, listeners_));
  }

  protected void onPut(FObject obj) {
    java.util.Iterator<DAOListener> iter = listeners_.iterator();

    while ( iter.hasNext() ) {
      DAOListener s = iter.next();
      s.put(obj);
    }
  }

  protected void onRemove(FObject obj) {
    java.util.Iterator<DAOListener> iter = listeners_.iterator();

    while ( iter.hasNext() ) {
      DAOListener s = iter.next();
      s.remove(obj);
    }
  }

  protected void onReset() {
    java.util.Iterator<DAOListener> iter = listeners_.iterator();

    while ( iter.hasNext() ) {
      DAOListener s = iter.next();
      s.reset();
    }
  }

  public FObject put(FObject obj) {
    return this.put_(this.getX(), obj);
  }

  public FObject remove(FObject obj) {
    return this.remove_(this.getX(), obj);
  }

  public void removeAll() {
    this.removeAll_(this.getX(), 0, this.MAX_SAFE_INTEGER, null, null);
  }

  public void removeAll_(X x, long skip, long limit, Comparator order, Predicate predicate) {
    this.select_(x, new RemoveSink(this), skip, limit, order, predicate);
  }

  protected Sink prepareSink(Sink s) {
    return s == null ? new ListSink() : s;
  }

  public Sink select() {
    return select(null);
  }

  public Sink select(Sink sink) {
    sink = prepareSink(sink);
    return this.select_(this.getX(), sink, 0, this.MAX_SAFE_INTEGER, null, null);
  }

  public FObject find(Object id) {
    return this.find_(this.getX(), id);
  }

  public void pipe(Sink sink) {
    this.pipe_(this.getX(), sink);
  }

  public DAO inX(X x) {
    ProxyDAO dao = new ProxyDAO(this);
    dao.setX(x);
    return dao;
  }

  public Object cmd_(X x, Object obj) {
    // TODO
    return null;
  }

  public Object cmd(Object obj) {
    return this.cmd_(this.getX(), obj);
  }
}
